nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/record/ResultSetRecordSetTest.java

```diff
@@ -283,41 +544,235 @@ private ResultSet givenResultSetForArrayThrowsException(boolean featureSupported
+        }
+    }
+
+    private void thenActualArrayElementTypesMatchExpected(Map<String, DataType> expectedTypes, RecordSchema actualSchema) {
+        for (RecordField recordField : actualSchema.getFields()) {
+            if (recordField.getDataType() instanceof ArrayDataType) {
+                ArrayDataType arrayType = (ArrayDataType) recordField.getDataType();
+                if (!arrayType.getElementType().equals(expectedTypes.get(recordField.getFieldName()))) {
+                    throw new AssertionError("Array element type for " + recordField.getFieldName()
+                            + " is not of expected type " + expectedTypes.get(recordField.getFieldName()).toString());
+                }
+            } else {
+                throw new AssertionError("RecordField " + recordField.getFieldName() + " is not instance of ArrayDataType");
+            }
+        }
+    }
+
+    private void thenAllDataTypesAreChoice(final List<RecordField> inputFields, final RecordSchema resultSchema) {
+        assertEquals("The number of input fields does not match the number of fields in the result schema.", inputFields.size(), resultSchema.getFieldCount());
+
+        DataType expectedType = getBroadestChoiceDataType();
+        for (int i = 0; i < inputFields.size(); ++i) {
+            assertEquals(expectedType, resultSchema.getField(i).getDataType());
+        }
+    }
+
+    private DataType getBroadestChoiceDataType() {
+        List<DataType> dataTypes = Stream.of(RecordFieldType.BIGINT, RecordFieldType.BOOLEAN, RecordFieldType.BYTE, RecordFieldType.CHAR, RecordFieldType.DATE,
+                RecordFieldType.DECIMAL, RecordFieldType.DOUBLE, RecordFieldType.FLOAT, RecordFieldType.INT, RecordFieldType.LONG, RecordFieldType.SHORT, RecordFieldType.STRING,
+                RecordFieldType.TIME, RecordFieldType.TIMESTAMP)
+                .map(RecordFieldType::getDataType)
+                .collect(Collectors.toList());
+        return RecordFieldType.CHOICE.getChoiceDataType(dataTypes);
+    }
+
+    private static class TestColumn {
+        private final int index; // Column indexing starts from 1, not 0.
+        private final String columnName;
+        private final int sqlType;
+        private final DataType recordFieldType;
+
+        public TestColumn(final int index, final String columnName, final int sqlType, final DataType recordFieldType) {
+            this.index = index;
+            this.columnName = columnName;
+            this.sqlType = sqlType;
+            this.recordFieldType = recordFieldType;
+        }
+
+        public int getIndex() {
+            return index;
+        }
+
+        public String getColumnName() {
+            return columnName;
+        }
+
+        public int getSqlType() {
+            return sqlType;
+        }
+
+        public DataType getRecordFieldType() {
+            return recordFieldType;
+        }
+    }
+
+    private static class SqlArrayDummy implements Array {
```

Comment:
Recommend renaming to `ResultSqlArray`:
```suggestion
    private static class ResultSqlArray implements Array {
```